x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/ttest/PairedTTestAggregator.java

```diff
@@ -0,0 +1,91 @@
+import org.elasticsearch.common.lease.Releasables;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.index.fielddata.SortedNumericDoubleValues;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.metrics.CompensatedSum;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.MultiValuesSource;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.analytics.ttest.TTestAggregationBuilder.A_FIELD;
+import static org.elasticsearch.xpack.analytics.ttest.TTestAggregationBuilder.B_FIELD;
+
+public class PairedTTestAggregator extends TTestAggregator<PairedTTestState> {
+    private TStatsBuilder statsBuilder;
+
+    PairedTTestAggregator(String name, MultiValuesSource.NumericMultiValuesSource valuesSources, int tails, DocValueFormat format,
+                          SearchContext context, Aggregator parent, List<PipelineAggregator> pipelineAggregators,
+                          Map<String, Object> metadata) throws IOException {
+        super(name, valuesSources, tails, format, context, parent, pipelineAggregators, metadata);
+        statsBuilder = new TStatsBuilder(context.bigArrays());
+    }
+
+    @Override
+    protected PairedTTestState getState(long bucket) {
+        return new PairedTTestState(statsBuilder.get(bucket), tails);
+    }
+
+    @Override
+    protected PairedTTestState getEmptyState() {
+        return new PairedTTestState(new TTestStats(0, 0, 0), tails);
+    }
+
+    @Override
+    protected long size() {
+        return statsBuilder.getSize();
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+                                                final LeafBucketCollector sub) throws IOException {
+        if (valuesSources == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        final BigArrays bigArrays = context.bigArrays();
+        final SortedNumericDoubleValues docAValues = valuesSources.getField(A_FIELD.getPreferredName(), ctx);
+        final SortedNumericDoubleValues docBValues = valuesSources.getField(B_FIELD.getPreferredName(), ctx);
+        final CompensatedSum compDiffSum = new CompensatedSum(0, 0);
+        final CompensatedSum compDiffSumOfSqr = new CompensatedSum(0, 0);
+
+        return new LeafBucketCollectorBase(sub, docAValues) {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                statsBuilder.grow(bigArrays, bucket + 1);
```

Comment:
I think we can move this inside the conditionals below?  Right now we'll end up growing the statsBuilder bigarray even if none of the documents end up satisfying the conditions (e.g. if we are unlucky and they all have only one of the two fields).

The size provided to `grow()` is the min size required, so it's ok to not call grow until we actually need a particular bucket ordinal (and then it will back-fill all the empty bucket ords essentially)